### PR TITLE
hcl2: handle unquoted undefined variables

### PR DESCRIPTION
**Pull request by notnoop**
This fixes a regression in https://github.com/hashicorp/nomad/pull/10326, to handle unquoted unknown variables.

The HCL job may contain unquoted undefined variable references without `${...}` wrapping, e.g. `value = meta.node_class`. In 1.0.4, this got parsed as `value = "${meta.node_class}"`.

This code performs a scan to find the relevant `${` and `}`, and only tries to find the closest ones with whitespace as the only separator.

### CODE REVIEW

File: jobspec2/types.config.go [hunk 1]
```diff
@@ -315,18 +314,32 @@ func (c *jobConfig) EvalContext() *hcl.EvalContext {
...
-				v += "}"
+
+			// find the start of inclusing "${..}" if it's inclused in one; otherwise, wrap it in one
+			quoted := false
+			for i := start - 1; i >= 1; i-- {
+				if body[i] == '{' && body[i-1] == '$' {
+					quoted = true
+					v = string(body[i-1:start]) + v
+					break
+				} else if body[i] != ' ' {
```
Comment (tgross) [inline jobspec2/types.config.go:325]: Should we being using [`unicode.IsSpace`](https://golang.org/pkg/unicode/#IsSpace) to catch other whitespace chars?

File: jobspec2/types.config.go [hunk 2]
```diff
@@ -315,18 +314,32 @@ func (c *jobConfig) EvalContext() *hcl.EvalContext {
...
 
 			v := string(body[start:end])
-			if i := bytes.IndexByte(body[end:], '}'); i != -1 {
-				v += string(body[end : end+i+1])
-			} else {
-				// fallback for unexpected cases
-				v += "}"
+
+			// find the start of inclusing "${..}" if it's inclused in one; otherwise, wrap it in one
+			quoted := false
```
Comment (tgross) [inline jobspec2/types.config.go:319]: I'm finding this a little hard to follow and maybe it's the name of this variable that's confusing me. Is this really only signifying "quoted" or is it doing double-duty for something like `isBracketed` as well?

Comment (notnoop) [reply jobspec2/types.config.go:319]: Bracketed is a better name - didn't know what to call it!  Thanks!

File: jobspec2/types.config.go (hunk 1 above)
Comment (notnoop) [reply jobspec2/types.config.go:325]: Testing emprically, only a plain space works. Trying with any other (e.g. `\t`, `\n`, etc) I get an error like the following:

```
input.hcl:2,23-24: Invalid character; This character is not used within the language.
input.hcl:2,23-24: Extra characters after interpolation expression; Expected a closing brace to end the interpolation expression, but found extra characters.
```

**Review by tgross — APPROVED** [review]
LGTM
